File: OsmAndWear/src/net/osmand/plus/widgets/style/CustomTypefaceSpan.java
package net.osmand.plus.widgets.style;

import android.graphics.Typeface;
import android.text.TextPaint;
import android.text.style.MetricAffectingSpan;

import androidx.annotation.NonNull;

public class CustomTypefaceSpan extends MetricAffectingSpan {

	private final Typeface typeface;

	public CustomTypefaceSpan(Typeface typeface) {
		this.typeface = typeface;
	}

	@Override
	public void updateMeasureState(@NonNull TextPaint p) {
		update(p);
	}

	@Override
	public void updateDrawState(TextPaint tp) {
		update(tp);
	}

	private void update(TextPaint tp) {
		Typeface old = tp.getTypeface();
		Typeface font = Typeface.create(typeface, old != null ? old.getStyle() : Typeface.NORMAL);
		tp.setTypeface(font);
	}
}

File: OsmAnd-telegram/src/net/osmand/telegram/ui/views/CustomTypefaceSpan.kt
package net.osmand.telegram.ui.views

import android.graphics.Typeface
import android.text.TextPaint
import android.text.style.MetricAffectingSpan

class CustomTypefaceSpan(val font: Typeface) : MetricAffectingSpan() {

	override fun updateMeasureState(textPaint: TextPaint) = update(textPaint)


	override fun updateDrawState(textPaint: TextPaint) = update(textPaint)

	private fun update(textPaint: TextPaint) {
		textPaint.apply {
			val old = typeface
			val oldStyle = old?.style ?: 0
			typeface = Typeface.create(font, oldStyle)
		}
	}
}